compiler/rustc_codegen_ssa/src/mir/block.rs

```diff
@@ -594,11 +594,19 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
                             );
                         }
 
+                        let type_err = || {
+                            panic!("trace inputs should be a struct");
+                        };
+
                         // Check it's the right type.
                         let local_decl = &self.mir.local_decls[ret_place.local];
                         match local_decl.ty.kind {
-                            ty::Tuple(_) => (),
```

Comment:
Why not allow both tuples and structs?

Reply:
I figured "tuple structs" would suffice. What do you think?

Reply:
A tuple is basically an anonymous struct. Allowing them both is less surprising I think and it can be nice in some cases.

Reply:
Yeah, I realised that they had the same ABI, but some part of me wants to be as restrictive as possible here.

@ptersilie thoughts?

Reply:
Why do we want to be restrictive?

Reply:
@bjorn3 by the way, why isn't a tuple a `ty::Adt` in rustc? There's a separate `ty::Tuple` variant.

Reply:
I'm with Edd that I would prefer the use of either but not both. While they are essentially the same we do end up in different paths during codegen and at least for now I find it easier if I only need to care about one type.

Reply:
OK, fair enough.

Reply:
But I agree that we could probably allow either in the future. Though there are not many compelling reasons to use a tuple over a struct.

Reply:
> @bjorn3 by the way, why isn't a tuple a ty::Adt in rustc? There's a separate ty::Tuple variant.

An `Adt` is only for types defined by the user using `struct`, `enum` or `union`. Unlike `Tuple` it has a `DefId` pointing to the definition. For tuples there is no `DefId` that can be used as it is not defined in any crate.

Reply:
I see. Thanks!